src/data/zsqrtd/basic.lean

```diff
@@ -571,4 +577,59 @@ instance : linear_ordered_semiring ℤ√d := by apply_instance
 instance : ordered_semiring ℤ√d        := by apply_instance
 
 end
+
+lemma norm_eq_zero {d : ℤ} (h_nonsquare : ∀ n : ℤ, d ≠ n*n) (a : ℤ√d) :
+  norm a = 0 ↔ a = 0 :=
+begin
+  refine ⟨λ ha, ext.mpr _, λ h, by rw [h, norm_zero]⟩,
+  delta norm at ha,
+  rw sub_eq_zero at ha,
+  by_cases h : 0 ≤ d,
+  { obtain ⟨d', rfl⟩ := int.eq_coe_of_zero_le h,
+    haveI : nonsquare d' := ⟨λ n h, h_nonsquare n $ by exact_mod_cast h⟩,
+    exact divides_sq_eq_zero_z ha, },
+  { push_neg at h,
+    suffices : a.re * a.re = 0,
+    { rw eq_zero_of_mul_self_eq_zero this at ha ⊢,
+      simpa only [true_and, or_self_right, zero_re, zero_im, eq_self_iff_true,
+        zero_eq_mul, mul_zero, mul_eq_zero, h.ne, false_or, or_self] using ha },
+    apply _root_.le_antisymm _ (mul_self_nonneg _),
+    rw [ha, mul_assoc],
+    exact mul_nonpos_of_nonpos_of_nonneg h.le (mul_self_nonneg _) }
+end
+
+variables {R : Type} [comm_ring R]
+
+/-- The unique `ring_hom` from `ℤ√d` to a ring `R`, constructed by replacing `√d` with the provided
+root. Conversely, this associates to every mapping `ℤ√d →+* R` a value of `√d` in `R`. -/
+@[simps]
+def lift {d : ℤ} : {r : R // r * r = ↑d} ≃ (ℤ√d →+* R) :=
```

Comment:
This is a stronger definition than before, now that it provides the map in the other direction too

Reply:
But is it just as usable? Wouldn't it be better to provide an `ext`-lemma for ring-homs out of `\Z\sqd`?

Reply:
>But is it just as usable?

I'd argue that `lift ⟨r, hr⟩` isn't really any less usable than `lift r hr`, and the advantage is you have all the `equiv` lemmas available too. I've already converted a large number of `lift`s in mathlib to this form, so I'd prefer to continue the trend.

An `ext` lemma is also a good idea though, presumably of the form `f sqrtd = g sqrtd \to f = g`?

Reply:
I added that lemma

Reply:
Thanks